File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> sub = new ArrayList();

        for (int num:nums) {
            System.out.println(num);
            int pos = bisectionLeft(sub, num);
            
            if (pos == sub.size()) {
                sub.add(num);
            } else {
                sub.set(pos, num);
            }
        }

        return sub.size();
    }
    public int bisectionLeft (List<Integer> list, int target) {
        int low = 0;
        int high = list.size() - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (list.get(mid) < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return low;
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        sub = []
        
        for num in nums:
            pos = self.bisectLeft(sub, num)
            if pos == len(sub):
                sub.append(num)
            else:
                sub[pos] = num
        
        return len(sub)

    # <bisect.bisect_left> = TC: O(log n)
    # finding the leftmost position in a sorted list
    # where a given target value should be inserted to maintain the list's sorted order
    def bisectLeft(self, list, target) -> int:
        low = 0
        high = len(list) - 1

        while low <= high :
            mid = int(low + (high - low) / 2) 

            print()

            if list[mid] < target:
                low = mid + 1
            else:
                high = mid - 1

        return low
